File: torchci/scripts/check_rockset.py
import difflib
import json
import os
import sys
from rockset import Client

API_KEY = os.environ["ROCKSET_API_KEY"]
API_SERVER = "https://api.rs2.usw2.rockset.com"

with open("./rockset/prodVersions.json") as f:
    versions = json.load(f)

rs = Client(api_server=API_SERVER, api_key=API_KEY)
failed = False
for query, version in versions.items():
    print(f"Checking that query: {query}:{version} matches your local checkout.")
    qlambda = rs.QueryLambda.retrieveByVersion(
        query, version=version, workspace="commons"
    )
    remote_query = qlambda["sql"]["query"]
    with open(f"./rockset/commons/__sql/{query}.sql") as f:
        local_query = f.read()
        if remote_query != local_query:
            failed = True
            print(f"::error::{query}:{version} does not match your local checkout.")
            diff = difflib.unified_diff(
                remote_query.splitlines(),
                local_query.splitlines(),
                fromfile="Rockset remote query",
                tofile="local checkout",
            )
            for line in diff:
                print(line)

if failed:
    exit(1)

File: torchci/scripts/checkRockset.mjs
import rockset from "@rockset/client";
import { promises as fs } from "fs";
import { diffLines } from "diff";
import "colors";

async function readJSON(path) {
  const rawData = await fs.readFile(path);
  return JSON.parse(rawData);
}

function printDiff(first, second) {
  diffLines(first, second).forEach((part) => {
    const color = part.added ? "green" : part.removed ? "red" : "grey";
    const prefix = part.added ? "+ " : part.removed ? "- " : "  ";
    part.value.split("\n").forEach((line) => {
      process.stderr.write(prefix[color] + line[color] + "\n");
    });
  });
}

async function checkQuery(client, workspace, queryName, version) {
  console.log(
    `Checking that ${workspace}.${queryName}:${version} is matches your local checkout`
  );
  let qLambda;
  const res = await client.queryLambdas.getQueryLambdaVersion(
    workspace,
    queryName,
    version
  );
  qLambda = res.data;
  let passesCheck = true;

  // Check that the query SQL matches the local checkout.
  const remoteQuery = qLambda.sql.query;
  const localQuery = await fs.readFile(
    `./rockset/${workspace}/__sql/${queryName}.sql`,
    "utf8"
  );
  if (remoteQuery.trim() !== localQuery.trim()) {
    console.log(
      `::error::${workspace}.${queryName}:${version} SQL does not match your local checkout.
       Run 'yarn node scripts/downloadQueryLambda.mjs ${workspace} ${queryName} ${version}' to overwrite your local checkout.`
    );
    printDiff(remoteQuery, localQuery);
    passesCheck = false;
  }

  // Check that the query config matches the local checkout.
  const localConfig = await readJSON(
    `./rockset/${workspace}/${queryName}.lambda.json`
  );

  const remoteParams = JSON.stringify(qLambda.sql.default_parameters, null, 2);
  const localParams = JSON.stringify(localConfig.default_parameters, null, 2);
  if (remoteParams !== localParams) {
    console.log(
      `::error::${workspace}.${queryName}:${version} config does not match your local checkout. Run 'yarn node scripts/downloadQueryLambda.mjs ${workspace} ${queryName} ${version}' to overwrite your local checkout.`
    );
    printDiff(remoteParams, localParams);
    passesCheck = false;
  }
  return passesCheck;
}

// NOTE: This is a special token with highly restricted access to Rockset. It
// can only be used to list resources. It is intended for use by the CI system only.
const client = rockset.default(
  "4td9ejgZWF4HERUmZufILSn9kQbixwX1FwR9CiJmTkWaXbVNi90eeyfgP7dO2BZP"
);

const checks = [];
const prodVersions = await readJSON("./rockset/prodVersions.json");
Object.keys(prodVersions).forEach((workspace) => {
  Object.entries(prodVersions[workspace]).forEach(([queryName, version]) =>
    checks.push(checkQuery(client, workspace, queryName, version))
  );
});

const checkStatuses = await Promise.all(checks);

if (checkStatuses.some((status) => status === false)) {
  process.exit(1);
}
